src/Servers/Kestrel/Core/test/SniOptionsSelectorTests.cs

```diff
@@ -0,0 +1,459 @@
+using Microsoft.AspNetCore.Server.Kestrel.Https.Internal;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
+{
+    public class SniOptionsSelectorTests
+    {
+        private static X509Certificate2 _x509Certificate2 = TestResources.GetTestCertificate();
+
+        [Fact]
+        public void PrefersExactMatchOverWildcardPrefixOverWildcardOnly()
+        {
+            var sniDictionary = new Dictionary<string, SniConfig>
+            {
+                {
+                    "www.example.org",
+                    new SniConfig
+                    {
+                        Certificate = new CertificateConfig
+                        {
+                            Path = "Exact"
+                        }
+                    }
+                },
+                {
+                    "*.example.org",
+                    new SniConfig
+                    {
+                        Certificate = new CertificateConfig
+                        {
+                            Path = "WildcardPrefix"
+                        }
+                    }
+                },
+                {
+                    "*",
+                    new SniConfig
+                    {
+                        Certificate = new CertificateConfig
+                        {
+                            Path = "WildcardOnly"
+                        }
+                    }
+                }
+            };
+
+            var mockCertificateConfigLoader = new MockCertificateConfigLoader();
+            var pathDictionary = mockCertificateConfigLoader.CertToPathDictionary;
+
+            var sniOptionsSelector = new SniOptionsSelector(
+                "TestEndpointName",
+                sniDictionary,
+                mockCertificateConfigLoader,
+                fallbackHttpsOptions: new HttpsConnectionAdapterOptions(),
+                fallbackHttpProtocols: HttpProtocols.Http1AndHttp2,
+                logger: Mock.Of<ILogger<HttpsConnectionMiddleware>>());
+
+            var wwwSubdomainOptions = sniOptionsSelector.GetOptions(new MockConnectionContext(), "www.example.org");
+            Assert.Equal("Exact", pathDictionary[wwwSubdomainOptions.ServerCertificate]);
+
+            var baSubdomainOptions = sniOptionsSelector.GetOptions(new MockConnectionContext(), "b.a.example.org");
+            Assert.Equal("WildcardPrefix", pathDictionary[baSubdomainOptions.ServerCertificate]);
+
+            var aSubdomainOptions = sniOptionsSelector.GetOptions(new MockConnectionContext(), "a.example.org");
+            Assert.Equal("WildcardPrefix", pathDictionary[aSubdomainOptions.ServerCertificate]);
+
+            // "*.example.org" is preferred over "*", but "*.example.org" doesn't match "example.org".
+            // REVIEW: Are we OK with "example.org" matching "*" instead of "*.example.org"? It feels annoying to me to have to configure example.org twice.
```

Comment:
Did you run this past Barry?

The rule in similar wildcard systems is that `foo` does not match `*.foo`, it only matches `*`.

Reply:
@blowdart 😄 

Not until now. I've written it to have the behavior you wanted where "\*.example.org" doesn't match "example.org" as annoying as it may be, because as I mention on the following line, the alternative would have "a.example.org" match "\*.a.example.org" before "\*.example.org", and that just seems wrong.

Reply:
So *.example.org should not match example.org, but it should match a.example.org, b.example.org, a.b.example.org and so on. That's how SANs in certificates work as well (almost) so it's not surprising.

A wildcard of just * should match everything IMO.

So what chris said :)